w3af/plugins/grep/vulners_scanner.py

```diff
@@ -0,0 +1,251 @@
+        ol.add(o)
+        return ol
+
+    def set_options(self, options_list):
+        """
+        This method sets all the options that are configured using the user
+        interface generated by the framework using the result of get_options().
+        :param options_list: A dictionary with the options for the plugin.
+        :return: No value is returned.
+        """
+        self._vulners_api_key = options_list['vulners_api_key'].get_value()
+
+    def update_vulners_rules(self):
+        # Get fresh rules from Vulners Github.
+        # Paranoid? Check gitlog and regexes.
+        # w3af grep plugins shouldn't (by definition) perform HTTP requests
+        # But in this case we're breaking that general rule to retrieve the
+        # DB at the beginning of the scan
+        # One more time after Retire.JS :)
+        try:
+            http_response = self._uri_opener.GET(self._vulners_rules_url,
+                                                 binary_response=True,
+                                                 respect_size_limit=False)
+        except Exception as e:
+            msg = 'Failed to download Vulners regex rules table: "%s"'
+            om.out.error(msg % e)
+            return
+
+        if http_response.get_code() != 200:
+            msg = ('Failed to download the Vulners regex rules table, unexpected'
+                   ' HTTP response code %s')
+            om.out.error(msg % http_response.get_code())
+            return
+
+        json_table = http_response.get_raw_body()
+        self.rules_table = json.loads(json_table)
+
+        # Adapt it for MultiRe structure [(regex,alias)] removing regex duplicated
+        regex_aliases = collections.defaultdict(list)
+        for software_name in self.rules_table:
+            regex_aliases[self.rules_table[software_name].get('regex')] += [software_name]
+        # Now create fast RE filter
+        # Using re.IGNORECASE because w3af is modifying headers when making RAW dump. Why so? Raw must be raw!
+        self._multi_re = MultiRE(((regex, regex_aliases.get(regex)) for regex in regex_aliases), re.IGNORECASE)
+
+    def get_vulners_api(self):
+        # Lazy import. Just not to make it in the __init__
+        # Other way API will try to communicate Vulners right at the w3af startup
+        if not self._vulners_api:
+            # Don't forget to setup Vulners API key if you have one
+            self._vulners_api = vulners.Vulners(api_key= self._vulners_api_key)
+        return self._vulners_api
+
+    def check_vulners(self, software_name, software_version, check_type):
+        vulnerabilities = {}
+        cached_result = self._vulnerability_cache.get((software_name, software_version, check_type))
+        if cached_result:
+            return cached_result
+        # Ask Vulners about vulnerabilities
+        if check_type == 'software':
+            vulnerabilities = self.get_vulners_api().softwareVulnerabilities(software_name, software_version)
```

Comment:
What happens if the user entered an invalid API key? Will this line raise an exception?

Reply:
It will. Yep, good idea to refactor this moment.